src/compas/datastructures/assembly/part.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+from collections import deque
+from functools import reduce
+
+from compas.geometry import Shape
+from compas.geometry import Frame
+from compas.geometry import multiply_matrices
+from compas.geometry import boolean_union_mesh_mesh
+from compas.geometry import boolean_difference_mesh_mesh
+from compas.geometry import boolean_intersection_mesh_mesh
+
+from ..datastructure import Datastructure
+from ..mesh import Mesh
+
+from .exceptions import FrameError
+from .exceptions import FeatureError
+
+
+class Part(Datastructure):
+    """A data structure for representing assembly parts."""
+
+    operations = {
+        'union': boolean_union_mesh_mesh,
+        'difference': boolean_difference_mesh_mesh,
+        'intersection': boolean_intersection_mesh_mesh
+    }
+
+    @property
+    def DATASCHEMA(self):
+        import schema
+        return schema.Schema({
+            "attributes": dict,
+            "key": int,
+            "frame": Frame,
+            "transformations": list,
+            "geometry": Shape,
+            "features": list
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        return 'assembly'
+
+    def __init__(self, name, **kwargs):
+        super(Part, self).__init__()
+        self._key = None
+        self._frame = None
+        self._transformations = deque()
+        self._geometry = None
+        self._features = []
+        self._shape = None
```

Comment:
The terms `geometry` and `shape` are confusingly close, but I can't think of a good alternative right now. `base_geometry` and `final_geometry` come to mind, but they are not really correct. Also, `base_geometry` might be the only geometry in a lot of cases, so it's a bit annoying to have to prefix it all the time

Reply:
is somewhat in line with the terminology in OCC

Reply:
the idea is that "shapes" are well-defined regular objects (like the shapes in `compas.geometry`). "geometry" is more general and can basically be anything, but cannot be expressed directly as a (named) shape. so `(base) shape + features => (general) geometry`